proto/spire/plugin/server/credentialcomposer/v1/credentialcomposer.proto

```diff
@@ -0,0 +1,179 @@
+
+    // Composes workload X509-SVIDs.
+    rpc ComposeWorkloadX509SVID(ComposeWorkloadX509SVIDRequest) returns (ComposeWorkloadX509SVIDResponse);
+
+    // Composes workload JWT-SVIDs.
+    rpc ComposeWorkloadJWTSVID(ComposeWorkloadJWTSVIDRequest) returns (ComposeWorkloadJWTSVIDResponse);
+}
+
+message ComposeServerX509CARequest {
+    // The default attributes for the server X509 CA.
+    X509CAAttributes attributes = 1;
+}
+
+message ComposeServerX509CAResponse {
+    // The attributes for the X509 CA. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509CAAttributes attributes = 1;
+}
+
+message ComposeServerX509SVIDRequest {
+    // The default attributes for the server X509-SVID.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeServerX509SVIDResponse {
+    // The attributes for the X509-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeAgentX509SVIDRequest {
+    // The default attributes for the server X509-SVID.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeAgentX509SVIDResponse {
+    // The attributes for the X509-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeWorkloadX509SVIDRequest {
+    // The default attributes for the X509-SVID.
+    X509SVIDAttributes attributes = 1;
+
+    // The SPIFFE ID of the workload.
+    string spiffe_id = 2;
+
+    // PKIX encoded public key of the workload.
+    bytes public_key = 3;
+}
+
+message ComposeWorkloadX509SVIDResponse {
+    // The attributes for the X509-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeWorkloadJWTSVIDRequest {
+    // The default attributes for the JWT-SVID.
+    JWTSVIDAttributes attributes = 1;
+
+    // The SPIFFE ID of the workload.
+    string spiffe_id = 2;
+}
+
+message ComposeWorkloadJWTSVIDResponse {
+    // The attributes for the JWT-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    JWTSVIDAttributes attributes = 1;
+}
+
+message X509CAAttributes {
+    // The subject of the X509 CA.
+    DistinguishedName subject = 1;
+
+    // Zero or more policy identifiers (OIDs) to apply to the CA.
+    repeated string policy_identifiers = 2;
+
+    // Zero or more extensions to apply to the X509 CA. These will override
+    // any extensions otherwise added by the other fields.
+    repeated X509Extension extra_extensions = 3;
+}
+
+message X509SVIDAttributes {
+    // The subject of the X509-SVID.
+    DistinguishedName subject = 1;
+
+    // Zero or more DNS SANs to apply to the X509-SVID.
+    repeated string dns_sans = 2;
```

Comment:
For the fields that can also be defined in the entry, what would be the logic to determine what wins?

Reply:
I imagined that we formulate these attributes from the entry, as we do today, and then invoke the plugin for the final determination. This order is important for some of the use cases.

Reply:
That sounds right. The only thing that I was thinking is if the protocol should dictate (in the comments/documentation) how that _must_ be handled, or if we see that plugin-dependent.

For example, through the `extra_extensions` field, the URI(s) from the SAN extension could be defined. I guess that the server would just set the URI depending on the SPIFFE ID of the entry, so if something different comes from this field it will just be ignored. I was wondering if we should document that kind of situation.

Reply:
Yeah, I can embellish the comments on extra_extensions some more. Namely, that it can override all otherwise specified extensions, but that if it produces an non SPIFFE spec-conformat certificate or attempts to change the SPIFFE ID, that the minting operation will fail.